src/main/java/redis/clients/jedis/JedisClusterConnectionHandler.java

```diff
@@ -34,12 +34,22 @@ public Jedis getConnectionFromNode(HostAndPort node) {
 
   private void initializeSlotsCache(Set<HostAndPort> startNodes, GenericObjectPoolConfig poolConfig) {
     for (HostAndPort hostAndPort : startNodes) {
-      Jedis jedis = new Jedis(hostAndPort.getHost(), hostAndPort.getPort());
+      Jedis jedis = null;
       try {
+        jedis = new Jedis(hostAndPort.getHost(), hostAndPort.getPort());
         cache.discoverClusterNodesAndSlots(jedis);
         break;
       } catch (JedisConnectionException e) {
-        // try next nodes
+        // close already created resource and try next nodes
+        for (JedisPool pool : cache.getNodes().values()) {
```

Comment:
@Spikhalskiy 
I don't think handler is responsible of destroying the pools in JedisClusterInfoCache. If we need to destroy, it would be better to handle in JedisClusterInfoCache, or at least it should be extracted to method of JedisClusterInfoCache.

Ideally we should destroy pool when calling `this.nodes.clear();`. Since we only clear nodes when clearing all states, we can extract this behavior into new method, maybe destroy() or reset().
Definitely it also applies to BinaryJedisCluster. Let cache destroy its pools by itself.

If you think it's beyond the boundary of pull request, we can merge as is and I'll submit new pull request.